source/parser.cc

```diff
@@ -0,0 +1,253 @@
+	AstNode *left = sv[0].get<ParseTreeNode>().get();
+	AstNode *right = sv[1u].get<ParseTreeNode>().get();
+	AstNode *el = nullptr;
+	if (sv.size() > 2u)
+		el = sv[2u].get<ParseTreeNode>().get();
+	left = new IfElseNode(left, right, el);
+	return left;
+}
+
+//block node
+AstNode *bind(const SemanticValues &sv) {
+	vector<AstNode*> vec;
+	int j = 0;
+	for (auto i = 0u; i < sv.size(); i+=1) {
+		vec.push_back(sv[i].get<ParseTreeNode>().get());
+		j++;
+	}
+	return new Block(vec, j);
+}
+
+//function definition node
+AstNode *funDef(const SemanticValues &sv) {
+	vector<AstNode*> vars;
+	AstNode* block;
+	for (auto i = 0u; i < sv.size()-1u; i+=1) {
+		vars.push_back(sv[i].get<ParseTreeNode>().get());
+	}
+	block = sv[sv.size()-1u].get<ParseTreeNode>().get();
+	return new FunDef(block, vars);
+
+}
+
+//function call node
+AstNode *funCall(const SemanticValues &sv) {
+	AstNode* ref = sv[0].get<ParseTreeNode>().get();
+	vector<AstNode*> vars;
+	for (auto i = 1u; i < sv.size(); i+=1)
+		vars.push_back(sv[i].get<ParseTreeNode>().get());
+
+	if (ref->to_string() == "print")
+		return new PrintNode(vars);
+	else
+		return new FunCall(vars, ref);
+	
+}
+
+//parse information for smurf
+void setup_ast_generation(parser &parser) {
+
+	parser["vDec"] = [](const SemanticValues &sv) {
+		AstNode *n = assign(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["assignment"] = [](const SemanticValues &sv) {
+		AstNode *n = assign(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["boolean_expression"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["arithmetic_expression"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["mult_term"] = [](const SemanticValues &sv) {
+		AstNode *n = bin_op(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["ifexpression"] = [](const SemanticValues &sv) {
+		AstNode *n = ifElse(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["identifier"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new VariableValue(sv.str()));
+	};
+
+	parser["variablereference"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new VariableValue(sv.str()));
+	};
+
+	parser["number"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new IntegerNode(atoi(sv.c_str())));
+	};
+
+	parser["add_op"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["mul_op"] = [](const SemanticValues &sv) {
+	        return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["relop"] = [](const SemanticValues &sv) {
+		return ParseTreeNode(new OpNode(sv.str()));
+	};
+
+	parser["block"] = [](const SemanticValues &sv) {
+		AstNode *n = bind(sv);
+		return ParseTreeNode(n);
+	};
+
+	parser["program"] = [](const SemanticValues &sv) {
+		AstNode *n = bind(sv);
+		ParseTreeNode val = ParseTreeNode(n);
+		Interpreter* I = new Interpreter();
+		val.get()->accept(I);
```

Comment:
You definitely have separate ASTs and visitors, but here you've build calling the visitor into the AST generation.

I'm not going to deduct for this, but the whole idea was to keep the two decoupled....